apps/aws-notebook/tf/main.tf

```diff
@@ -0,0 +1,123 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">=4"
+    }
+  }
+}
+
+
+provider "aws" {
+  default_tags {
+    tags = {
+      Sandbox = data.external.env.result.sandbox_name
+    }
+  }
+}
+
+variable "ecs_cluster_name" {
+  type        = string
+  description = "The name of ECS cluster"
+  nullable    = false
+  validation {
+    condition     = length(var.ecs_cluster_name) > 0
+    error_message = "The ecs_cluster_name value must greater than 0."
+  }
+}
+
+
+variable "subnet_id" {
+  type        = string
+  description = "subnet id for ECS service"
+  nullable    = false
+  validation {
+    condition     = length(var.subnet_id) > 0
+    error_message = "The subnet_id value must greater than 0."
+  }
+}
+
+variable "security_groups" {
+  type     = string
+  nullable = true
+}
+
+variable "service_launch_type" {
+  type    = string
+  default = "FARGATE"
+}
+
+variable "ssh_public_key" {
+  type    = string
+  default = ""
+}
+
+variable "task_image" {
+  type     = string
+  nullable = false
+  validation {
+    condition     = length(var.task_image) > 0
+    error_message = "The task_image value must greater than 0."
+  }
+}
+
+data "aws_ecs_cluster" "cluster" {
+  cluster_name = var.ecs_cluster_name
+}
+
+data "external" "env" {
+  program = ["${path.module}/env.sh"]
+}
+
+data "aws_iam_role" "ecs_task_execution_role" {
+  name = "ecsTaskExecutionRole"
+}
+
+# Create a task definition 
+resource "aws_ecs_task_definition" "notebook" {
+  family                   = "notebook_${data.external.env.result.sandbox_id}"
+  requires_compatibilities = ["FARGATE", "EC2"]
+  network_mode             = "awsvpc"
+  cpu                      = 256
+  memory                   = 512
+  runtime_platform {
+    operating_system_family = "LINUX"
+    cpu_architecture        = "X86_64"
+  }
+  task_role_arn      = data.aws_iam_role.ecs_task_execution_role.arn
+  execution_role_arn = data.aws_iam_role.ecs_task_execution_role.arn
+  container_definitions = jsonencode([
+    {
+      name      = "notebook"
+      image     = var.task_image
+      essential = true
+      environment = [
+        {
+          name  = "PUBLIC_KEY"
+          value = var.ssh_public_key
+        }
+      ]
+      portMappings = [
+        {
+          containerPort = 22
+          hostPort      = 22
+        }
+      ]
+    }
+  ])
+}
+
+# Create a service
+resource "aws_ecs_service" "notebook" {
+  launch_type         = var.service_launch_type
+  task_definition     = aws_ecs_task_definition.notebook.arn
+  name                = "notebook_${data.external.env.result.sandbox_id}"
+  cluster             = data.aws_ecs_cluster.cluster.id
+  scheduling_strategy = "REPLICA"
+  desired_count       = 1
+  network_configuration {
+    subnets          = [var.subnet_id]
+    assign_public_ip = false
+    security_groups  = split(",", var.security_groups)
+  }
+}
```

Comment:
I'd recommend add an output using terraform template which invokes `get-task-ip.sh` to generate the output. So the lifecycle is fully managed by terraform.

Reply:
Did you mean save the ip to `terraform output` by  run get-task-ip.sh in `data external ` ?